core/shared/src/main/scala/kantan/csv/HeaderDecoder.scala

```diff
@@ -36,18 +36,18 @@ object HeaderDecoder extends GeneratedHeaderDecoders {
   /** Summons an implicit instance of [[HeaderDecoder]] if one can be found, fails compilation otherwise. */
   def apply[A](implicit ev: HeaderDecoder[A]): HeaderDecoder[A] = macro imp.summon[HeaderDecoder[A]]
 
-  // Horrible code that relies on mutability and assumes that `map` contains the right values, indexed from 0 to length.
-  private[csv] def decoderFromMap(map: Map[String, Int], header: Seq[String]): DecodeResult[Array[Int]] = {
-    val mapping = Array.fill(map.size)(-1)
-
-    header.zipWithIndex.foreach {
-      case (name, i) =>
-        map.get(name).foreach(a => mapping(a) = i)
+  private def determineRowMappings(requiredHeader: Seq[String], csvHeader: Seq[String]): DecodeResult[Seq[Int]] = {
+    val indicesOrMissingHeaders: Seq[Either[String, Int]] = requiredHeader.map { header =>
```

Comment:
The only issue I have with this is that you're doing two passes over something that could conceivably be done in one. The code might end up faster and slightly more readable if you folded directly on `requiredHeader`, accumulating on a `(Seq[String], Seq[Int])`, and then deciding whether to return a success of a failure depending on whether the list of errors is empty.